src/java.base/share/classes/com/sun/crypto/provider/ChaCha20Cipher.java

```diff
@@ -753,6 +769,120 @@ protected int engineDoFinal(byte[] in, int inOfs, int inLen, byte[] out,
+        throws ShortBufferException {
+        return bufferCrypt(input, output, false);
+    }
+
+    /*
+     * Optimized version of bufferCrypt from CipherSpi.java.  Direct
+     * ByteBuffers send to the engine code.
+     */
+    private int bufferCrypt(ByteBuffer input, ByteBuffer output,
+        boolean isUpdate) throws ShortBufferException {
+        if ((input == null) || (output == null)) {
+            throw new NullPointerException
+                ("Input and output buffers must not be null");
+        }
+        int inPos = input.position();
+        int inLimit = input.limit();
+        int inLen = inLimit - inPos;
+        if (isUpdate && (inLen == 0)) {
+            return 0;
+        }
+        int outLenNeeded = engine.getOutputSize(inLen, !isUpdate);
+
+        if (output.remaining() < outLenNeeded) {
+            throw new ShortBufferException("Need at least " + outLenNeeded
+                + " bytes of space in output buffer");
+        }
+
+        int total = 0;
+
+        // Check if input bytebuffer is heap-backed
+        if (input.hasArray()) {
+            byte[] inArray = input.array();
+            int inOfs = input.arrayOffset() + inPos;
+
+            byte[] outArray;
+            // Check if output bytebuffer is heap-backed
+            if (output.hasArray()) {
+                outArray = output.array();
+                int outPos = output.position();
+                int outOfs = output.arrayOffset() + outPos;
+
+                // check array address and offsets and use temp output buffer
+                // if output offset is larger than input offset and
+                // falls within the range of input data
+                boolean useTempOut = false;
+                if (inArray == outArray &&
+                    ((inOfs < outOfs) && (outOfs < inOfs + inLen))) {
+                    useTempOut = true;
+                    outArray = new byte[outLenNeeded];
+                    outOfs = 0;
+                }
+                try {
+                    if (isUpdate) {
+                        total = engine.doUpdate(inArray, inOfs, inLen, outArray, outOfs);
+                    } else {
+                        total = engine.doFinal(inArray, inOfs, inLen, outArray, outOfs);
+                    }
+                } catch (BadPaddingException | KeyException e) {
+                    throw new ProviderException(e);
+                }
+                if (useTempOut) {
+                    output.put(outArray, outOfs, total);
+                } else {
+                    // adjust output position manually
+                    output.position(outPos + total);
+                }
+            } else { // if output is direct
+                if (isUpdate) {
+                    outArray = engineUpdate(inArray, inOfs, inLen);
+                } else {
+                    try {
+                        outArray = engineDoFinal(inArray, inOfs, inLen);
+                    } catch (BadPaddingException e) {
```

Comment:
Should we really be hiding the `BadPaddingException` here and in other places in this method? `AEADBadTagException` extends `BadPaddingException`, and I'm pretty sure we don't want to hide it.

Reply:
CipherSpi:engineUpdate() only throws ShortBufferException.  That forces many exceptions into a ProviderException with extends RuntimeException.  The BadPaddingException doesn't matter because CC20 doesn't have padding.  AEADTagException should be specified and thrown, not wrapped.